### ISSUE TITLE
[help]Can't grab Bluetooth packets on my device

### ISSUE THREAD

**Issue by yyy33** [LuigiVampa92/XLogger#1]
![Screenshot_20230304-231528-454](https://user-images.githubusercontent.com/51548147/222914223-ae8b1f92-090b-440c-86c7-9fe7c1abd449.png)

**Comment by yyy33** [comment]
![Screenshot_20230304-231837](https://user-images.githubusercontent.com/51548147/222914246-673c3458-1ec9-469f-b9c4-7c1385a260a0.png)

**Comment by LuigiVampa92** [comment]
Hi, thanks for your interest to the project! Are Magisk and LSPosed are properly installed on your device? Please check that it is so, and that XLogger module is activated for the app that you want to grab communication from.

Also please note that for now there are some limitations with catching a Bluetooth communication that I have mentioned on a readme page. Only Bluettooth LE communications that most devices use nowadays are intercepted, but not classic old EDR (BluetoothSocket). Also you have to wait some time and do not close the app you are grabbing communication from for some time (up to a couple of dozens of seconds in the worst case) to make sure that it will be saved properly.

If this won't help I'll try to think about something else

**Comment by yyy33** [comment]
> Hi, thanks for your interest to the project! Are Magisk and LSPosed are properly installed on your device? Please check that it is so, and that XLogger module is activated for the app that you want to grab communication from.
> 
> Also please note that for now there are some limitations with catching a Bluetooth communication that I have mentioned on a readme page. Only Bluettooth LE communications that most devices use nowadays are intercepted, but not classic old EDR (BluetoothSocket). Also you have to wait some time and do not close the app you are grabbing communication from for some time (up to a couple of dozens of seconds in the worst case) to make sure that it will be saved properly.
> 
> If this won't help I'll try to think about something else

Yes I have read wiki before using the module and have the correct magisk and lsposed environment and configuration. The problem I have now is I can t be sure what protocol the target software uses ble or edr. If the software uses edr, this is expected according to the instructions on wiki, if the software uses ble, that is that the package is not properly captured by the module

**Comment by LuigiVampa92** [comment]
That's very possible. The code is quite raw, I haven't tested it widely, I managed to capture BLE on a couple of my devices, but of course it might not work in 100% cases. Can you please make a debug build and check what's dropped there?

You can make a debug build simply by:
```
./gradlew assembleDebug
```
And then check logcat with the log tag:
```
adb logcat | grep XLOGGER
```
Or you can message me in telegram, my nickname there is also, and I'll make and send you a debug build

**Comment by yyy33** [comment]
> That's very possible. The code is quite raw, I haven't tested it widely, I managed to capture BLE on a couple of my devices, but of course it might not work in 100% cases. Can you please make a debug build and check what's dropped there?
> 
> You can make a debug build simply by:
> 
> ```
> ./gradlew assembleDebug
> ```
> 
> And then check logcat with the log tag:
> 
> ```
> adb logcat | grep XLOGGER
> ```
> 
> Or you can message me in telegram, my nickname there is also luigivampa92, and I'll make and send you a debug build

I'm not an Android developer, so I need your help with my build. I sent you a message on Telegraph, did you get it (I'm using Telegraph for the first time and not too familiar with the operation)

**Comment by LuigiVampa92** [comment]
Sorry, changed confidentiality settings. Can you try again please?

**Comment by yyy33** [comment]
> Sorry, changed confidentiality settings. Can you try again please?

Still the same，Can you send me a message, the name is yyy33

**Comment by yyy33** [comment]
> Sorry, changed confidentiality settings. Can you try again please?

sorry ,my name is @yyy33_cool

**Comment by yyy33** [comment]
> Sorry, changed confidentiality settings. Can you try again please?

I still can't send a message to no, probably because I just registered

**Comment by yyy33** [comment]
> Sorry, changed confidentiality settings. Can you try again please?

I still can't send you a message, probably because I just registered

**Comment by LuigiVampa92** [comment]
Deleted comments that contained the personal data

**Comment by LuigiVampa92** [comment]
For history:

The problem was in incorrect hooking of gatt callback classes. I covered the case when the callback class was hooked directly, but not the case when callback class was extended and then instantiated. While child classes ARE technically the instances of parent classes for JVM, It's not the same for Xposed framework. So I have added a code that scans classes in DEX file, checks if there are any non-abstract children of gatt callbacks and if there are some - hook those classes along with the original callback class.

The new version with this fix has been published here on releases page, and on the LSPosed modules repository.

Problem had been solved, so I close this issue.
